packages/libsql/src/lib/LibSqlClient.ts

```diff
@@ -0,0 +1,203 @@
+      })
+      await trx.commit()
+      return { type: "Written", position: 0n, snapshot_etag: nextEtag }
+    } catch (err: any) {
+      await trx.rollback()
+      throw err
+    }
+  }
+}
+
+export class LibSqlReader {
+  constructor(private readonly client: Client) {}
+
+  async readLastEventWithType(streamName: string, eventType: string) {
+    const result = await this.client.execute({
+      sql: "select * from messages where stream_name = ? and type = ? order by position desc limit 1",
+      args: [streamName, eventType],
+    })
+    if (result.rows.length === 0) return null
+    return Parse.row(result.rows[0])
+  }
+
+  async readLastEvent(streamName: string, eventType?: string) {
+    if (eventType != null) return this.readLastEventWithType(streamName, eventType)
+    const result = await this.client.execute({
+      sql: "select * from messages where stream_name = ? order by position desc limit 1",
+      args: [streamName],
+    })
+    if (result.rows.length === 0) return null
+    return Parse.row(result.rows[0])
+  }
+
+  async readStream(
+    streamName: string,
+    fromPosition: bigint,
+    batchSize: number,
+  ): Promise<ITimelineEvent<Format>[]> {
+    const result = await this.client.execute({
+      sql: "select * from messages where stream_name = ? and position >= ? order by position limit ?",
+      args: [streamName, fromPosition.toString(), batchSize],
+    })
+    return result.rows.map(Parse.row)
+  }
+
+  async readSnapshot(streamName: string): Promise<[string, ITimelineEvent] | undefined> {
+    const result = await this.client.execute({
+      sql: "select * from snapshots where stream_name = ?",
+      args: [streamName],
+    })
+    if (result.rows.length === 0) return
+    const row = result.rows[0]
+    return [Parse.string(row.etag), Parse.row(row)]
+  }
+}
+
+type DbRow = {
+  id: string
+  time: string
+  type: string
+  data?: string
+  meta?: string
+  position: string
+}
+
+namespace Parse {
+  export const string = (s: unknown): string => {
+    if (typeof s === "string") return s
+    throw new Error(`Expected string, got ${typeof s}`)
+  }
+  export const bigint = (s: unknown): bigint => {
+    switch (typeof s) {
+      case "string":
+      case "number":
+        return BigInt(s)
+      default:
+        throw new Error(`Expected string, got ${typeof s}`)
+    }
+  }
+
+  export const date = (s: unknown): Date => {
+    if (typeof s === "string") return new Date(s)
+    throw new Error(`Expected string, got ${typeof s}`)
+  }
+
+  export const nullable =
+    <T>(parser: (s: unknown) => T) =>
+    (s: unknown): T | undefined => {
+      if (s == null) return
+      return parser(s)
+    }
+
+  export const row = (r: any): ITimelineEvent<Format> => {
+    const row = r as DbRow
+    const data = nullable(string)(row.data)
+    const meta = nullable(string)(row.meta)
+    const type = string(row.type)
+    const time = date(row.time)
+    const index = bigint(row.position)
```

Comment:
but you used directly as a version in readVersion
I feel position is fine as a role name for arguments, but an unambiguous `index` is best for the schema. Though minIndex and maxIndex might be even better for that usage too (am thinking I want to go through all the stores and remove usage of the position name now)